### DISCUSSION TITLE
PHP7.4 only | Unhandled "Uninitialized" in serialization

### DISCUSSION THREAD

**Discussion by kohlerdominik** [laravel/framework#33637] [General] [answered]
- Laravel Version: 7.21.0
- PHP Version: 7.4.4

### Description:
The with PHP7.4 introduced Typed Properties V2 have also added a new state.

> If a typed property does not have a default value, no implicit null default value is implied (even if the property is nullable). Instead, the property is considered to be uninitialized. 

https://wiki.php.net/rfc/typed_properties_v2#uninitialized_and_unset_properties

If a typed property, which is initalized in the `handle()`-method, is used in a Job using `SerializesModels` trait, the job will break on Serialization with the Exception

> Typed property [class]::[property] must not be accessed before initialization

### Steps To Reproduce:

- Create a new Job with a typed property, which you do not initalize (see example below)
- Queue the Job
- You will get `Typed property [class]::[property] must not be accessed before initialization`-Exception

```php
class ProcessPodcast implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    // this property can be serialized, as it is initalized in the constructor (before serialize-deserialize)
    protected Podcast $podcast;

    // but when it tries to serialize this property to push the job to the queue, 
    // you will get the exception, as the handle-method has not run yet
    protected AudioProcessor $processor;

    public function __construct(Podcast $podcast)
    {
        $this->podcast = $podcast;
    }

    public function handle(AudioProcessor $processor)
    {
        $this->processor = $processor;

       // here's some very complex stuff which uses the processor in various methods, that's why it's made accessible as class-property.
    }
}
```

### Suggested Solution:

In PHP7.4, the corresponding method [ReflectionProperty::isInitialized](https://www.php.net/manual/en/reflectionproperty.isinitialized.php) was introduced. So for >= 7.4 we can just check this before accessing the the the property, for example in [SerializesModels:129](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Queue/SerializesModels.php#L129). But I do not yet know how to resolve it in a backwards compatible way.

### Workaround:

Anyone finding this and in need of a workaround:
Create a nullable typed property instead and set a default value, like:
```php
protected ?AudioProcessor $processor = null;
```

**Comment by driesvints** [comment]
We only support model serialization in jobs sorry. I'll move this to discussions as a feature request. Feel free to attempt a PR if you like. Thanks.

**Reply by GrahamCampbell** [reply]
This can happen with models too, actually. See https://github.com/laravel/framework/pull/33644.

**Reply by judgej** [reply]
Did this ever get fixed? I'm getting the same error deserializing a job with a simple string typed property. PHP deserialization is happy to deserialize such an object, but Laravel - I assume - is doing some additional reflection magic here? Making the property nullable and giving it a null default value fixes it, but feels very much a workaround.

Using PHP 8.1 Laravel 8.83

**Comment by kohlerdominik** [comment]
Thanks for the quick response. When I find some time I'll try to work out a PR with a backwards-compatible fix for this.

**Comment by GrahamCampbell** [comment, marked as answer]
This problem will be fixed in the releases on Tuesday.

**Reply by kohlerdominik** [reply]
Thanks for addressing my issue. 

I think this might be unrelated enough to create a new issue, but I'll ask here anyway:
While doing this, i found a workaround for not stored models in our code, too.

Basically the issue is that you run in an error when trying to serialize a model which is not yet stored. It might be edge-case, but in our usecase we use temporary model instances, so instances that are never actually stored to the database.

```php
class ProcessPodcast implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    // at serialization time, this is a fresh model instance, but it's not stored to database. So obviously it has no Key yet.
    // trying to get the key for serialization will fail.
    protected Podcast $podcast;

    public function __construct(Podcast $podcast)
    {
        $this->podcast = new Podcast();
    }

    public function handle(AudioProcessor $processor)
    {
        $this->podcast->save();
    }
}
```

In our fix we check the `$podcasts->exists` attribute. If the model `exists`, we can serialize by "SerializesModels". But if this check fails, we serialize it like it was no `QueueableEntity`.
